pkg/registry/memory/nse_memory_server.go

```diff
@@ -28,28 +28,55 @@ import (
 
 type memoryNetworkServeRegistry struct {
 	storage *Storage
-	nsmName string
 }
 
 func (m *memoryNetworkServeRegistry) RegisterNSE(ctx context.Context, registration *registry.NSERegistration) (*registry.NSERegistration, error) {
+	nseRegistration, err := m.processRegistration(registration)
+	if err != nil {
+		return nseRegistration, err
+	}
+
+	return next.NetworkServiceRegistryServer(ctx).RegisterNSE(ctx, registration)
+}
+
+func (m *memoryNetworkServeRegistry) processRegistration(registration *registry.NSERegistration) (*registry.NSERegistration, error) {
 	if registration == nil {
 		return nil, errors.New("can not register nil registration")
 	}
 	registration.NetworkServiceEndpoint.State = "RUNNING"
-	registration.NetworkServiceEndpoint.NetworkServiceManagerName = m.nsmName
-	nsm, ok := m.storage.NetworkServiceManagers.Load(m.nsmName)
-	if !ok {
-		return nil, errors.New("network service manager is not found")
-	}
-	registration.NetworkServiceManager = nsm
 
 	m.storage.NetworkServiceEndpoints.Store(registration.NetworkServiceEndpoint.Name, registration.NetworkServiceEndpoint)
 	m.storage.NetworkServices.Store(registration.NetworkService.Name, registration.NetworkService)
-	return next.NetworkServiceRegistryServer(ctx).RegisterNSE(ctx, registration)
+	return registration, nil
+}
+
+type nsmBulkRegServer struct {
+	registry.NetworkServiceRegistry_BulkRegisterNSEServer
+	server registry.NetworkServiceRegistry_BulkRegisterNSEServer
+	m      *memoryNetworkServeRegistry
+}
+
+func (ts *nsmBulkRegServer) Send(reg *registry.NSERegistration) error {
+	nseRegistration, err := ts.m.processRegistration(reg)
+	if err != nil {
+		return err
+	}
+	return ts.server.Send(nseRegistration)
+}
+func (ts *nsmBulkRegServer) Recv() (*registry.NSERegistration, error) {
+	return ts.server.Recv()
+}
+
+func (ts *nsmBulkRegServer) Context() context.Context {
+	return ts.server.Context()
 }
 
 func (m *memoryNetworkServeRegistry) BulkRegisterNSE(s registry.NetworkServiceRegistry_BulkRegisterNSEServer) error {
-	return next.NetworkServiceRegistryServer(s.Context()).BulkRegisterNSE(s)
+	bulkRegS := &nsmBulkRegServer{
+		m:      m,
+		server: s,
+	}
```

Comment:
I would suggest to remove field `server` from structure nsmBulkRegServer.
```suggestion
	bulkRegS := &nsmBulkRegServer{
		m:      m,
		NetworkServiceRegistry_BulkRegisterNSEServer: s,
	}
```

Reply:
Thanks, will changes to custom actually.